Objects/longobject.c

```diff
@@ -3611,24 +3617,43 @@ long_richcompare(PyObject *self, PyObject *other, int op)
     Py_RETURN_RICHCOMPARE(result, 0, op);
 }
 
+void
+_PyLong_ExactDealloc(PyObject *self)
+{
+    if (_PyLong_IsCompact((PyLongObject *)self)) {
+        _Py_FREELIST_FREE(ints, self, PyObject_Free);
```

Comment:
I just realised that we didnt check for if it's a smallint cached value here. It should be rejected there

Reply:
Nevermind. It's not needed because (theoretically) the refcount of a smallint never hits 0 due to it being immortal, so this code will never be called on it. I'll just add an assert to be safe.

Reply:
This also means my comment about the small ints was wrong. Could you remove that please?

Reply:
The comment has been adapted. If the refcount can never become zero, then the code in `long_dealloc` checking for small ints can be removed (perhaps also with an assert). Since this is something unrelated to the freelists I will make a separate PR for that and rebase this one.